shark-graph/src/main/java/shark/internal/HprofInMemoryIndex.kt

```diff
@@ -72,60 +68,148 @@ internal class HprofInMemoryIndex private constructor(
+    }
+    array = primitiveArrayIndex[objectId]
+    if (array != null) {
+      return IndexedPrimitiveArray(
+          array.readTruncatedLong(positionSize), PrimitiveType.values()[array.readByte().toInt()]
+      )
+    }
+    return null
+  }
+
   fun objectIdIsIndexed(objectId: Long): Boolean {
-    return objectIndex[objectId] != null
+    if (classIndex[objectId] != null) {
+      return true
+    }
+    if (instanceIndex[objectId] != null) {
+      return true
+    }
+    if (objectArrayIndex[objectId] != null) {
+      return true
+    }
+    if (primitiveArrayIndex[objectId] != null) {
+      return true
+    }
+    return false
   }
 
   private class Builder(
+    longIdentifiers: Boolean,
+    fileLength: Long,
+    classCount: Int,
+    instanceCount: Int,
+    objectArrayCount: Int,
+    primitiveArrayCount: Int,
     private val indexedGcRootsTypes: Set<KClass<out GcRoot>>
   ) : OnHprofRecordListener {
+
+    private val identifierSize = if (longIdentifiers) 8 else 4
+    private val positionSize = byteSizeForUnsigned(fileLength)
+
     /**
      * Map of string id to string
      * This currently keeps all the hprof strings that we could care about: class names,
      * static field names and instance fields names
      */
     // TODO Replacing with a radix trie reversed into a sparse array of long to trie leaf could save
-    // memory.
-    // Another option is to switch back to reading from the file system as necessary, and keep a much
-    // smaller cache for strings we need during shortest path (those are for exclusions)
+    // memory. Can be stored as 3 arrays: array of keys, array of values which are indexes into
+    // a large array of string bytes. Each "entry" consists of a size, the index of the previous
+    // segment and then the segment content.
+
     private val hprofStringCache = LongObjectScatterMap<String>()
 
     /**
      * class id to string id
      */
     private val classNames = LongLongScatterMap()
 
-    /**
-     * Object id to [IndexedObject].
-     * The id can be for classes instances, classes, object arrays and primitive arrays
-     */
-    private val objectIndex =
-      LongObjectScatterMap<IndexedObject>()
+    private val classIndex = UnsortedByteEntries(
+        bytesPerValue = positionSize + identifierSize + 4, longIdentifiers = longIdentifiers,
+        initialCapacity = classCount
+    )
+    private val instanceIndex = UnsortedByteEntries(
+        bytesPerValue = positionSize + identifierSize, longIdentifiers = longIdentifiers,
+        initialCapacity = instanceCount
+    )
+    private val objectArrayIndex = UnsortedByteEntries(
+        bytesPerValue = positionSize + identifierSize, longIdentifiers = longIdentifiers,
+        initialCapacity = objectArrayCount
+    )
+    private val primitiveArrayIndex = UnsortedByteEntries(
+        bytesPerValue = positionSize + 1, longIdentifiers = longIdentifiers, initialCapacity = primitiveArrayCount
+    )
```

Comment:
Nitpick: putting parameters on separate lines will make it a bit easier to read; right now the alignment is inconsistent

Reply:
yeah sometimes I get lazy because the formatter isn't doing this for me. Will fix